### ISSUE TITLE
Incorrect Padding applied when using buffers

### ISSUE THREAD

**Issue by peteroreil** [maxdavidson/structly#104]
When using two buffers of 8 byte length, the incorrect offset is being applied.
If a float64 is used in place of once of these 64bit buffers, the correct padding is applied.

To reproduce:

```
"use strict";
let structly = require('structly');


let createConverter = structly.createConverter;

let struct = structly.struct;
let bool = structly.bool;
let string = structly.string;
let uint32le = structly.uint32le;
let int32le = structly.int32le;
let uint16le = structly.uint16le;
let int16le = structly.int16le;
let int8 = structly.int8;
let uint8 = structly.uint8;
let array = structly.array;
let float32 = structly.float32;
let float64le = structly.float64le;
let buffer = structly.buffer;

//define the structure of our packet
let MiddlePayLoad = struct({
    mpl1: bool,
    mpl2: string(1, 'utf8') //char
});

let Payload = struct({
    pl1: uint8,
    pl2: uint16le,
    pl3: uint32le,
    pl4: buffer(8),
    mpl: MiddlePayLoad,
    pl5: int8,
    pl6: int16le,
    pl7: int32le,
    pl8: buffer(8)
});

let SuperStruct = struct({
    d1: string(1, 'utf8'),
    d2: int16le,
    d3: int32le,
    d4: string(1, 'utf8'),
    pl: Payload
});

let max64Bit =  new Buffer([0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff]);
let min64Bit =  new Buffer([0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]);

let data = {
    d1: 'E',
    d2: 4567,
    d3: 11111,
    d4: 'R',
    pl: {
        pl1: 256,
        pl2: 65535,
        pl3: 4294967295,
        pl4: max64Bit,
        
        mpl: {
            mpl1: true,
            mpl2: 'M'
        },
        
        pl5: -128,
        pl6: -32768,
        pl7: -2147483648,
        pl8: min64Bit
    }
};




let converter = createConverter(SuperStruct);
let bytes = converter.encode(data);

console.log(bytes.toString('hex'));
console.log('buffer length', bytes.length);
```

This produces the following output

```
4500d711672b0000520000000000ffffffffffffffffffffffffffff014d800000800000000000800000000000000000
buffer length 48
```

If we modify the Payload Struct above and replace the `buffer` with a `float64le`  and modify the data we get the expected buffer length and correct offsets.

```
let Payload = struct({
    pl1: uint8,
    pl2: uint16le,
    pl3: uint32le,
    pl4: buffer(8),
    mpl: MiddlePayLoad,
    pl5: int8,
    pl6: int16le,
    pl7: int32le,
    pl8: float64le
});

...

let data = {
    d1: 'E',
    d2: 4567,
    d3: 11111,
    d4: 'R',
    pl: {
        pl1: 256,
        pl2: 65535,
        pl3: 4294967295,
        pl4: max64Bit,
        
        mpl: {
            mpl1: true,
            mpl2: 'M'
        },
        
        pl5: -128,
        pl6: -32768,
        pl7: -2147483648,
        pl8: 1234
    }
};

```

Output is as follows:

```
4500d711672b000052000000000000000000ffffffffffffffffffffffffffff014d80000080000000000080000000000000000000489340
buffer length 56

```

I think the offending code is in the `struct` function in `schemas.js` where the byteOffset is not being correctly calculated for a `buffer` object.

If I modify the `buffer` `byteAlignment` to equal its length, it works as expected

```
/** Create a buffer schema */
function buffer(length) {
    if (typeof length !== 'number') {
        throw new TypeError('You must specify a length of the buffer!');
    }
    return {
        tag: 7 /* Buffer */,
        version: exports.SCHEMA_VERSION,
        byteLength: length,
        byteAlignment: length
    };
}
exports.buffer = buffer;
```

**Comment by maxdavidson** [comment]
A "buffer" has no inherit alignment, so it's basically equivalent to `array(uint8, n)`, which has an alignment of  1 due to the `sizeof(uint8)` being 1. Setting `byteAlignment` to `length` is not correct, since it must be a power of 2, e.g. 1, 2, 4, or 8. You do raise a valid point though that the alignment should be configurable, so that you should be able to do `buffer(8, alignof(float64))`.

**Comment by maxdavidson** [comment]
In the meantime, this hack should work:
```js
function alignedBuffer(length, byteAlignment = 1) {
  if (byteAlignment <= 0) {
    throw new RangeError('Byte alignment must be positive');
  }
  if (Math.log2(byteAlignment) % 1 !== 0) {
    throw new TypeError('Byte alignment must be a power of 2');
  }
  return Object.assign(buffer(length), { byteAlignment });
}

const myBuffer = alignedBuffer(8, 8);
```

**Comment by maxdavidson** [comment]
Fixed in `v0.8.0` with `buffer(8, alignof(float64))`
